include/quicly/loss.h

```diff
@@ -265,7 +265,9 @@ inline void quicly_loss_update_alarm(quicly_loss_t *r, int64_t now, int64_t last
     do {                                                                                                                           \
         int64_t _t = (t);                                                                                                          \
         if (is_after_send) {                                                                                                       \
-            assert(now < _t);                                                                                                      \
+            /*mask this assert, but don't modify the logic                                                                         \
+            n case is_after_send is true don't update _t                                                                           \
+              assert(now < _t);*/                                                                                                  \
```

Comment:
@odedk the `assert()` expects `t_` to be in the future.

```
4760     if (ret == 0) {
4761         /* update timers, start / stop delivery rate estimator */
4762         if (conn->application == NULL || conn->application->super.unacked_count == 0)
4763             conn->egress.send_ack_at = INT64_MAX; /* we have sent ACKs for every epoch (or before address validation) */
4764         int can_send_stream_data = scheduler_can_send(conn);
4765         update_send_alarm(conn, can_send_stream_data, 1);
```
`is_after_send` is set to 1 only in `do_send()` in the branch above. The condition that triggers crashes is this one:

```
317     SET_ALARM(last_retransmittable_sent_at + alarm_duration);
```

It looks like `alarm_duration` is always positive. So the expectation is that `last_retransmittable_sent_at >= now`.

The corresponding variables are `conn->egress.last_retransmittable_sent_at` and `conn->stash.now`, and the later is used for setting the former. So most likely (I did some checks though) that `conn→stash.now` is OK.

What I guess happens is that `conn→egress.last_retransmittable_sent_at` is actually not updated in one of the paths leading to that `if (ret == 0)` branch above. And so `update_send_alarm(…, 1)` is called with `last_retransmittable_sent_at` "from the past" and alarm_duration doesn't compensate this (and it likely shouldn't).

I did see one path that seemed to be suspicious (it looked like it didn't update `last_retransmittable_sent_at` but ret was 0) but I don't recall it immediately now.

If we remove `assert()`, most likely we must make sure that `t_` (`r→alarm_at`) is not in the past. For example, there are checks like this in the code:

```
 765         assert(conn->stash.now < conn->egress.loss.alarm_at);
```

Reply:
@katzoded report it upstream? Maybe they'll prefer fixing this specific path or just changing `SET_ALARM()`. We'll merge this fix for now.